### PR TITLE
Add more cortex queries

### PR DESCRIPTION
**Pull request by Anko59**

### CODE REVIEW

**Comment by Kamforka** [comment]
I'll take a closer look but on the high level it seems fine!
Maybe one thing that is not ideal is the simple return types. 
Probably we can define some more advanced type hints for those?

File: thehive4py/endpoints/cortex.py [hunk 1]
```diff
@@ -27,3 +27,26 @@ def create_responder_action(
                 "responderId": responder_id,
             },
         )
+
+    def list_analyzers(self) -> dict:
```
Comment (Kamforka) [inline thehive4py/endpoints/cortex.py:31]: So it looks like we miss a query parameter `range` based on the [api docs](https://docs.strangebee.com/thehive/api-docs/#tag/Cortex/operation/List%20Analyzers)

Additionally I'd also refine the output type hint as according to the docs it should be a list of analyzers.

For that please define the module `thehive4py/types/cortex.py` and add something like: 

```python
# thehive4py/types/cortex.py
from typing import List, TypedDict


class OutputAnalyzertRequired(TypedDict):
    id: str
    name: str
    version: str
    description: str


class OutputAnalyzer(OutputAnalyzerRequired, total=False):
    dataTypeList: List[str]
    cortexIds: List[str]
```


```suggestion
    def list_analyzers(self, range: str) -> List[OutputAnalyzer]:
```

File: thehive4py/endpoints/cortex.py [hunk 2]
```diff
@@ -27,3 +27,26 @@ def create_responder_action(
...
+        return self._session.make_request("GET", path="/api/connector/cortex/analyzer")
+
+    def list_analyzers_by_type(self, data_type: str) -> dict:
```
Comment (Kamforka) [inline thehive4py/endpoints/cortex.py:34]: Based on the above findings and suggestion please revisit all these methods too and align them with the docs if they are not aligned already.

**Review by Kamforka — CHANGES_REQUESTED** [review]

**Comment by Anko59** [comment]
Thanks for the review,

I applied the changes that you suggested. I added cortex types for the objects returned by the API and added the range parameter to the list_analyzers function. I did not find any other missing parameter.

I used python 3.9+ typing (`dict` and `list` instead of `Dict` and `List`). It was already like this in the cortex.py, though I noticed most of the repo still uses pre 3.9 typing.

To better match the documentation, "responder action" and "analyzer job" could be renamed "cortex action" and "cortex job", but I think the first ones are easier to understand.

File: thehive4py/types/cortex.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
+# thehive4py/types/cortex.py
```
Comment (Kamforka) [inline thehive4py/types/cortex.py:1]: remove this comment please, I just added it in the example as clarity :)

File: thehive4py/endpoints/cortex.py [hunk 4]
```diff
@@ -17,7 +24,7 @@ def create_analyzer_job(
 
     def create_responder_action(
         self, object_id: str, object_type: str, responder_id: str
```
Comment (Kamforka) [inline thehive4py/endpoints/cortex.py:26]: When checking the corresponding [docs](https://docs.strangebee.com/thehive/api-docs/#tag/Cortex/operation/Create%20an%20action) I can see that it supports more parameters for this endpoint, what about adding an input type hint too? E.g.:

```python
class InputResponderActionRequired(TypedDict):
    responderId: str
    objectType: str
    objectId: str

<...>
```

and then:

```suggestion
        self, action: InputResponderAction
```

File: thehive4py/endpoints/cortex.py [hunk 5]
```diff
@@ -1,10 +1,17 @@
...
+    OutputAnalyzerJob,
+    OutputResponder,
+    OutputResponderAction,
+)
+from typing import Optional
 
 
 class CortexEndpoint(EndpointBase):
     def create_analyzer_job(
         self, cortex_id: str, analyzer_id: str, observable_id: str
```
Comment (Kamforka) [inline thehive4py/endpoints/cortex.py:13]: same suggestion as for the responder action one

File: thehive4py/endpoints/cortex.py [hunk 6]
```diff
@@ -27,3 +34,31 @@ def create_responder_action(
                 "responderId": responder_id,
             },
         )
+
+    def list_analyzers(self, range: Optional[str] = None) -> list[OutputAnalyzer]:
+        payload = {"range": range} if range else None
+        return self._session.make_request(
+            "GET", path="/api/connector/cortex/analyzer", json=payload
+        )
```
Comment (Kamforka) [inline thehive4py/endpoints/cortex.py:42]: As I see in the [docs](https://docs.strangebee.com/thehive/api-docs/#tag/Cortex/operation/List%20Analyzers) `range` should be a query parameter, e.g.:

```suggestion
        params = {"range": range}
        return self._session.make_request(
            "GET", path="/api/connector/cortex/analyzer", params=params
        )
```

And as far as I remember you don't really need the if-else check in this case as when a query parameter is `None` the `requests` library will just ignore it.

File: thehive4py/endpoints/cortex.py [hunk 7]
```diff
@@ -27,3 +34,31 @@ def create_responder_action(
...
+
+    def list_analyzers_by_type(self, data_type: str) -> list[OutputAnalyzer]:
+        return self._session.make_request(
+            "GET", path=f"/api/connector/cortex/analyzer/type/{data_type}"
+        )
+
+    def get_analyzer(self, object_id: str) -> OutputAnalyzer:
```
Comment (Kamforka) [inline thehive4py/endpoints/cortex.py:49]: typo:
```suggestion
    def get_analyzer(self, analyzer_id: str) -> OutputAnalyzer:
```

**Review by Kamforka — CHANGES_REQUESTED** [review]

**Comment by Kamforka** [comment]
> I applied the changes that you suggested. I added cortex types for the objects returned by the API and added the range parameter to the list_analyzers function. I did not find any other missing parameter.

Nice, great job so far!

> I used python 3.9+ typing (`dict` and `list` instead of `Dict` and `List`). It was already like this in the cortex.py, though I noticed most of the repo still uses pre 3.9 typing.

Unfortunately we still support 3.8, so can you please fall back to `typing.List` and `typing.Dict` where necessary? I plan to drop support 2025Q1.

> To better match the documentation, "responder action" and "analyzer job" could be renamed "cortex action" and "cortex job", but I think the first ones are easier to understand.

I agree with the naming convention, more to the point than cortex action and job!

My last point to this PR would be the missing integration tests. Unfortunately we don't have a Cortex instance configured yet, so we cannot test all the endpoints, but probably we can add some dummy test for the `list_analyzers` and `list_responders` methods, would you like to give it a try?

File: tests/test_cortex_endpoint.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
+from thehive4py.client import TheHiveApi
+
+
+class TestCortexEndpoint:
+    def test_list_analyzers(self, thehive: TheHiveApi):
+        analyzers = thehive.cortex.list_analyzers()
+        assert isinstance(analyzers, list)
+        for analyzer in analyzers:
+            assert "name" in analyzer.keys()
+            assert "id" in analyzer.keys()
```
Comment (Kamforka) [inline tests/test_cortex_endpoint.py:10]: Let's keep it even simpler:

```suggestion
        assert analyzers == []
```

File: tests/test_cortex_endpoint.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+    def test_list_analyzers_by_type(self, thehive: TheHiveApi):
+        data_type = "mail"
+        analyzers = thehive.cortex.list_analyzers_by_type(data_type=data_type)
+        assert isinstance(analyzers, list)
+        for analyzer in analyzers:
+            assert "name" in analyzer.keys()
+            assert "id" in analyzer.keys()
```
Comment (Kamforka) [inline tests/test_cortex_endpoint.py:18]: Same goes for this one:
```suggestion
        assert analyzers == []
```

File: tests/test_cortex_endpoint.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+    def test_list_and_get_analyzers(self, thehive: TheHiveApi):
+        analyzers = thehive.cortex.list_analyzers()
+        for analyzer in analyzers:
+            _analyzer = thehive.cortex.get_analyzer(analyzer["id"])
+            assert _analyzer == analyzer
```
Comment (Kamforka) [inline tests/test_cortex_endpoint.py:24]: This is unfortunately not doing anything additional as the `analyzers` list should be empty without a Cortex instance attached, so probably best to remove it for now

File: tests/test_cortex_endpoint.py [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+        thehive.case.create(case={"title": "my first case", "description": "..."})
+        entity_type = "case"
+        entity_id = thehive.case.find()[0]["_id"]
+        responders = thehive.cortex.list_responders(
+            entity_type=entity_type, entity_id=entity_id
+        )
+        assert isinstance(responders, list)
+        for responder in responders:
+            assert "name" in responder.keys()
+            assert "id" in responder.keys()
```
Comment (Kamforka) [inline tests/test_cortex_endpoint.py:36]: You can also reuse the `test_case` fixture to simplify this:
```suggestion
     def test_list_responders(self, thehive: TheHiveApi, test_case: OutputCase):
        responders = thehive.cortex.list_responders(
            entity_type="case", entity_id=test_case["_id"]
        )
        assert responders == []
```

**Review by Kamforka — CHANGES_REQUESTED** [review]

**Comment by Kamforka** [comment]
> I implemented the changes you suggested. The tests are really dummy, as there are no responders or analyzers on the test instance, some of the test code is never reached. I don't know how to add responders or analyzers using the API. I will try to document myself.

First of all thank you very much for the changes, great job just as before!!!

To reply to your dummy test concern, unfortunately this is the best we can do at the moment as we don't yet have a live Cortex instance behind the integrator instance of TheHive, so we will always get back empty responder and analyzer lists, but that's okay for now. TheHive cannot create/define analyzers nor responders.

Additionally I added some simplification ideas to your tests, if you implement those I think we are done with this PR.

File: tests/test_cortex_endpoint.py [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+    def test_list_analyzers_by_type(self, thehive: TheHiveApi):
+        data_type = "mail"
+        analyzers = thehive.cortex.list_analyzers_by_type(data_type=data_type)
+        assert analyzers == []
+
+    def test_list_responders(self, thehive: TheHiveApi, test_case: OutputCase):
+        responders = thehive.cortex.list_responders(
+            entity_type="case", entity_id=test_case["_id"]
+        )
+        assert isinstance(responders, list)
```
Comment (Kamforka) [inline tests/test_cortex_endpoint.py:18]: Let's stick to the convention of the tests above:
```suggestion
        assert responders == []
```

**Review by Kamforka — CHANGES_REQUESTED** [review]

**Review by Kamforka — APPROVED** [review]

**Comment by Kamforka** [comment]
Thank you for your coop, nice one!

**Pull request merged**